google/cloud/security/notifier/pipelines/email_pipeline.py

```diff
@@ -0,0 +1,133 @@
+
+        Returns:
+            The output filename for the violations json.
+        """
+        now_utc = datetime.utcnow()
+        output_timestamp = now_utc.strftime(OUTPUT_TIMESTAMP_FMT)
+        output_filename = VIOLATIONS_JSON_FMT.format(self.resource,
+                                                     self.cycle_timestamp,
+                                                     output_timestamp)
+        return output_filename
+
+    def _write_temp_attachment(self):
+        """Write the attachment to a temp file.
+
+        Returns:
+            The output filename for the violations json just written.
+        """
+        # Make attachment
+        output_file_name = self._get_output_filename()
+        output_file_path = '{}/{}'.format(TEMP_DIR, output_file_name)
+        with open(output_file_path, 'w+') as f:
+            f.write(parser.json_stringify(self.violations))
+        return output_file_name
+
+    def _make_attachment(self):
+        """Create the attachment object.
+
+        Returns:
+            The attachment object.
+        """
+        output_file_name = self._write_temp_attachment()
+        attachment = self.mail_util.create_attachment(
+            file_location='{}/{}'.format(TEMP_DIR, output_file_name),
+            content_type='text/json',
+            filename=output_file_name,
+            disposition='attachment',
+            content_id='Violations'
+        )
+
+        return attachment
+
+    def _make_content(self):
+        """Create the email content.
+
+        Returns:
+            A tuple containing the email subject and the content
+        """
+        timestamp = datetime.strptime(
+            self.cycle_timestamp, '%Y%m%dT%H%M%SZ')
+        pretty_timestamp = timestamp.strftime("%d %B %Y - %H:%M:%S")
+        email_content = self.mail_util.render_from_template(
+            'notification_summary.jinja', {
+                'scan_date':  pretty_timestamp,
+                'resource': self.resource,
+                'violation_errors': self.violations,
+            })
+
+        email_subject = 'Forseti Violations {} - {}'.format(
+            pretty_timestamp, self.resource)
+        return email_subject, email_content
+
+    def _send(self, **kwargs):
+        """Send a summary email of the scan.
+
+        Args:
+            subject: Email subject
+            conetent: Email content
+            attachment: Attachment object
+        """
+        subject = kwargs.get('subject')
+        content = kwargs.get('content')
+        attachment = kwargs.get('attachment')
+
+        self.mail_util.send(email_sender=self.pipeline_config['sender'],
+                            email_recipient=self.pipeline_config['recipient'],
+                            email_subject=subject,
+                            email_content=content,
+                            content_type='text/html',
+                            attachment=attachment)
+
+    def run(self):
+        """Run the email pipeline"""
```

Comment:
If we implement the abstractmethod ```_compose()```, then these steps will formalize into the below.  And ```compose()``` would be where we can build the attachment, subject, and content into at lease a ```notification``` dictionary.

```python
notification = self._compose()
self._send(notification)

notification = {
  'subject': 'foo',
  'content': 'bar',
  'attachment': baz
}
```